java/statestore/src/test/java/sleeper/statestore/StateStoreProviderWithSizeTest.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.statestore;
+
+import org.junit.jupiter.api.Test;
+
+import sleeper.configuration.properties.instance.InstanceProperties;
+import sleeper.configuration.properties.table.TableProperties;
+import sleeper.core.schema.Schema;
+import sleeper.core.statestore.StateStore;
+import sleeper.core.statestore.inmemory.StateStoreTestHelper;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static sleeper.configuration.properties.InstancePropertiesTestHelper.createTestInstanceProperties;
+import static sleeper.configuration.properties.instance.CommonProperty.STATESTORE_PROVIDER_CACHE_SIZE;
+import static sleeper.configuration.properties.table.TablePropertiesTestHelper.createTestTableProperties;
+import static sleeper.configuration.properties.table.TableProperty.TABLE_ID;
+import static sleeper.configuration.properties.table.TableProperty.TABLE_NAME;
+import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;
+
+public class StateStoreProviderWithSizeTest {
+    private final InstanceProperties instanceProperties = createTestInstanceProperties();
+    private final Schema schema = schemaWithKey("key");
+    private final Map<String, StateStore> tableIdToStateStore = new HashMap<>();
+    private final List<String> tablesLoaded = new ArrayList<>();
+
+    @Test
+    void shouldCacheStateStore() {
+        // Given
+        TableProperties table = createStateStore("test-table-id", "test-table");
+
+        // When
+        Stream<StateStore> stateStores = providerGettingStores(table, table, table);
+
+        // Then
+        assertThat(stateStores).containsExactly(
+                stateStore(table),
+                stateStore(table),
+                stateStore(table));
+        assertThat(tablesLoaded).containsExactly("test-table-id");
+    }
+
+    @Test
+    void shouldRemoveOldestCacheEntryWhenLimitHasBeenReached() {
+        // Given
+        instanceProperties.setNumber(STATESTORE_PROVIDER_CACHE_SIZE, 2);
+        TableProperties table1 = createStateStore("table-id-1", "test-table-1");
+        TableProperties table2 = createStateStore("table-id-2", "test-table-2");
+        TableProperties table3 = createStateStore("table-id-3", "test-table-3");
+
+        // When
+        Stream<StateStore> stateStores = providerGettingStores(
+                table1,
+                table2,
+                // Cache size has been reached, oldest state store will be removed from cache
+                table3,
+                table1);
+
+        // Then
+        assertThat(stateStores).containsExactly(
+                stateStore(table1),
+                stateStore(table2),
+                stateStore(table3),
+                stateStore(table1));
```

Comment:
It looks like this would still pass if the state store was actually null as it comes back from the HashMap. This is pretty close to duplicating the production code. In most tests where we do this we perform some operations on the state store, so it can't be null.

One option would be to have separate variables for the TableProperties and StateStore object so we don't need to rely on the HashMap for the assertions.